### ISSUE TITLE
Falcon7b demo gives repeating outputs for certain input prompts

### ISSUE THREAD

**Issue by skhorasganiTT** [tenstorrent-metal/tt-metal#7273]
For some very specific input prompts, the Falcon7b demo gives output responses with indefinitely repeating tokens. It is not certain whether this is due to a bug or due to slightly incorrect predictions. Based on a preliminary investigation, it seems like the latter.

**Example 1:**
```
pytest --disable-warnings -q -s --input-method=cli --cli-input="Tell me a joke" models/demos/falcon7b/demo/demo.py
```
TT-metal output: `I'm a joke. I'm a joke. I'm a joke. ...`
Huggingface output: `I once told a joke to a friend, but it didn't work. He just looked at me and said, 'That's not funny.'`

**Example 2:**
```
pytest --disable-warnings -q -s --input-method=cli --cli-input="write a short poem about Shanghai in English" models/demos/falcon7b/demo/demo.py
```
TT-metal output:
![image](https://github.com/tenstorrent-metal/tt-metal/assets/159165450/f20419cf-e050-408b-acb3-932e0af35e21)

Huggingface output:
![image](https://github.com/tenstorrent-metal/tt-metal/assets/159165450/8965863a-814b-404e-b890-8a7e952530f7)

CC @uaydonat

**Comment by skhorasganiTT** [comment]
Addressed in #7315 by replacing the greedy decoding strategy with top-k-top-p sampling which improves quality of generated tokens and introduces stochasticity

**Comment by davorchap** [comment]
> Addressed in #7315 by replacing the greedy decoding strategy with top-k-top-p sampling which improves quality of generated tokens and introduces stochasticity

@skhorasganiTT you guys have a plan to run top-k/top-p all in device without the need to go to host?

**Comment by skhorasganiTT** [comment]
@davorchap We don't currently have a plan to do that, but eventually we should. Here are some numbers from the Falcon7b demo (decode with batch 32) we can use to determine priority:
Average embedding + forward pass time: 0.105 s
Average top-k top-p sampling time: 0.0017 s
Average greedy (argmax) time: 0.0012 s

Based on these, the overhead does not seem too significant, however if we want to pursue it, the relevant ops we'd need are top-k, multinomial sampling, and top-p filtering (sort + cumulative sum - see top_k_top_p_filtering in https://huggingface.co/transformers/v3.2.0/_modules/transformers/generation_utils.html)
